### PR TITLE
Make `Route`, `Group` and `MatchingResult` dispatcher independent

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | -

Extracted from #196

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 - New #195: Add debug collector for `yiisoft/yii-debug` (@xepozz)
 - Chg #207: Replace two `RouteCollectorInterface` methods `addRoute()` and `addGroup()` to single `addRoute()` (@vjik)
 - Enh #202: Add support for `psr/http-message` version `^2.0` (@vjik)
+- Chg #222: Make `Route`, `Group` and `MatchingResult` dispatcher independent (@rustamwin, @vjik)
```
Comment (rustamwin) [inline CHANGELOG.md:8]:
```suggestion
- Chg #222: Make `Route`, `Group` and `MatchingResult` dispatcher-independent (@rustamwin, @vjik)
```

File: UPGRADE.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+# Upgrading Instructions for Yii Router
+
+This file contains the upgrade notes for the Yii Router.
+These notes highlight changes that could break your application when you upgrade it from one major version to another.
+
+## 4.0.0
+
+In this release classes `Route`, `Group` and `MatchingResult` are made dispatcher independent. Now you don't can inject
```
Comment (rustamwin) [inline UPGRADE.md:8]:
```suggestion
In this release classes `Route`, `Group` and `MatchingResult` are made dispatcher-independent. Now you don't can inject
```

File: src/Route.php [hunk 3]
```diff
@@ -297,31 +282,29 @@ public function __debugInfo()
...
-        }
-
-        /** @var mixed $definition */
-        foreach ($this->middlewareDefinitions as $index => $definition) {
-            if (in_array($definition, $this->disabledMiddlewareDefinitions, true)) {
-                unset($this->middlewareDefinitions[$index]);
-            }
-        }
+        $this->enabledMiddlewares = array_values(
+            array_filter(
```
Comment (rustamwin) [inline src/Route.php:302]: `array_filter` function is faster than `foreach`?

**Review by rustamwin — CHANGES_REQUESTED** [review]
Need to invalidate `enabledMiddlewares` in functions that updates `middlewares` property.

**Comment by vjik** [comment]
> Need to invalidate `enabledMiddlewares` in functions that updates `middlewares` property.

Done.

File: src/Route.php (hunk 3 above)
Comment (vjik) [reply src/Route.php]: Seems, foreach faster (I'm search in web, not make own bench). Returned `foreach`.

**Review by rustamwin — APPROVED** [review]
There is a line that can be covered.

**Pull request merged**